pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+type DrainOptions struct {
+	NodeName           string
+	Force              bool
+	GracePeriodSeconds int
+}
+
+const (
+	drain_long = `Drain node in preparation for maintenance.
+
+The given node will be marked unschedulable to prevent new pods from arriving.
+Then all running pods will be removed from the machine.  If there are any
+unreplicated pods, no pods will be touched without --force.
+
+When you are ready to put the node back into service, use kubectl undrain.`
+	drain_example = `# Drain node "foo", even if there are unreplicated pods on it.
+$ kubectl drain foo --force
+
+# As above, but don't delete unreplicated pods, and use a grace period of 15 minutes.
+$ kubectl drain foo --grace-period=900
+`
+)
+
+func NewCmdDrain(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "drain NODE",
+		Short:   "Drain nodes in preparation for maintenance",
+		Long:    drain_long,
+		Example: drain_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(
+				RunDrain(f, out, cmd, args, options))
+		},
+	}
+	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. Ignored if negative.")
+	return cmd
+}
+
+func RunDrain(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *DrainOptions) error {
+	if len(args) != 1 {
+		return cmdutil.UsageError(cmd, fmt.Sprintf("USAGE: %s [flags]", cmd.Use))
+	}
+	options.NodeName = args[0]
+
+	mapper, _ := f.Object()
+	c, err := f.Client()
+	if err != nil {
+		return err
+	}
+	if err = maybeLameNode(mapper, c, out, options); err != nil {
+		return err
+	}
+	pods, err := c.Pods(api.NamespaceAll).List(nil, fields.Set{"spec.nodeName": options.NodeName}.AsSelector())
+	if err != nil {
+		return err
+	}
+	if err = scanForReplicationControllers(c, out, options, pods); err != nil {
```

Comment:
Yeah, these are good points. Maybe we should punt on this feature for now -- just blindly remove all pods, at least for the first version. I think even removing the daemon pods is OK because they should be re-created (deleting the mirror pod won't delete the static pod that's actually on the node) so it's basically a no-op.